### ISSUE TITLE
Custom repository instantiated apart from it's declaration does not work

### ISSUE THREAD

**Issue by skalashnyk** [wovalle/fireorm#69]
Hi there.

Custom repository, instantiated somewhere (in another file) apart from it's declaration, does not work. I mean, his custom methods (not inherited from `BaseFirestoreRepository`) don't work (got error "CustomeRepository.customMethod is not a funcytion")

This happens because `@CustomRepository()` decorator is not invoked in this case.
Also, we have `GetCustomRepository()` function, but it is not used (even in examples). How can we use it?

**Comment by wovalle** [comment]
Unfortunately this is a limitation of the decorators. If the file (with the decorator) is not required/invoked in runtime, the decorator won't be triggered (and in your case, the custom repository is not being stored).

I had exactly this limitation [in another project](https://github.com/alveflo/tsmediator/issues/12) and we ended up implementing a method so you have to import all your files, in fireorm it doesn't make a lot of sense the same strategy.

In TypeORM they solve this by requiring the user to specify a path in [the configuration file](https://typeorm.io/#/using-ormconfig) and then they require all the files on runtime. It works for them since they have a cli so they have a need to navigate into the filesystem, fireorm doesn't.

TLDR:  I don't know any way

**Comment by wovalle** [comment]
About the `GetCustomRepository()` function, it returns the custom repository registered for a collection or `undefined` if it doesn't have any (unlike `GetRepository()` which would return the base repository).

About the documentation, that's something I'll work fully in the coming weeks. I know the documentation is far from ideal, but that'll change :)

EDIT: Issue closed by mistake.

**Comment by theochampion** [comment]
Hey guys, I'm running into the same issue with a custom repository. From my understanding, as I import it in `community.ts`, the decorator should be evaluated but it doesnt seems to be the case.

The code i'm using:

`repositories/community_repository.ts`
```typescript
@CustomRepository(CommunityModel)
export default class CustomCommunityRepository extends BaseFirestoreRepository<
  CommunityModel
> {
  async createWithMembership(
    name: string,
  ): Promise<void> {
    // custom operations
  }
}
```

Then if I import it and try to access the custom repo in another file like so:

`app/community.ts`
```typescript

import { getRepository, getCustomRepository } from 'fireorm';

// repositories 
import CustomCommunityRepository from '../repositories/community_repository';

// models
import CommunityModel from '../models/community_model';

const communityRepo = getCustomRepository(
      CommunityModel
    ) as CustomCommunityRepository;

const res =  communityRepo.createWithMembership(
      name,
      description,
      type,
      tag_map,
      links,
      created_by
    );
```

I get an exception indicating that the decorator was not evaluated:

```
Error: 'Community' does not have a custom repository.
```

Any ideas? @wovalle 

Cheers! ✨

**Comment by wovalle** [comment]
Hey! I don't see anything bad with your code. Can you reproduce it in a minimal repo and post it here?

**Comment by Shashi-GS** [comment]
Hi, I am facing same problem still "'User' does not have a custom repository.". Is this problem solved, my code is exactly same as Jul 3rd comment?

**Comment by wovalle** [comment]
Hey @Shashi-GS, I could not replicate the Issue. Can you reproduce it and create a minimal repo? Thanks!

**Comment by Shashi-GS** [comment]
Thank you for the quick response. Here is the sample repo for your reference (https://github.com/Shashi-GS/fireorm_sample). 
Note : 
- create a user collection with email field in firestore database.
- add firestore serviceAccountKey details into the file.

1. npm install
2. npm run build
3. npm start
4. Access rest API post method
http://localhost:5000/test 

body as raw input 
{
  "query": "{test { email }}"
}

You will get error as 
"message": "this.userRepo.getUserByEmail is not a function",

This means it is not identifying as custom repo. If I replace the AuthService line from "return this.userRepo.getUserByEmail();" to
"return this.userRepo.findOne()" it will work properly. 

Please check and let me know how I can solve this issue.

**Comment by skalashnyk** [comment]
@Shashi-GS , I'm getting another error on running your example:
```
[Nest] 61618   - 09/23/2020, 10:58:13 AM   [ExceptionsHandler] Cannot create property 'message' on string 'No Access' +46673ms
TypeError: Cannot create property 'message' on string 'No Access'
    at Object.<anonymous> (/Users/skalashnyk/IdeaProjects/fireorm_sample/node_modules/apollo-server-core/dist/runHttpQuery.js:63:27)
    at Generator.next (<anonymous>)
    at fulfilled (/Users/skalashnyk/IdeaProjects/fireorm_sample/node_modules/apollo-server-core/dist/runHttpQuery.js:5:58)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)

```

It seems, we should pass some token...
You didn't describe this in your instructions above.

**Comment by Shashi-GS** [comment]
@skalashnyk ,

 Could you please tell me in what stage you are facing this issue ?

rest API invocation screenshot for your reference.
![image](https://user-images.githubusercontent.com/17082069/93986636-38585b80-fda4-11ea-87bf-71cf0b2a6eb0.png)

**Comment by skalashnyk** [comment]
@Shashi-GS , when trying to make an http-request (the same you did in tour screenshot).
Could you please check your "Headers" tab?

It seems this caused by these lines:
https://github.com/Shashi-GS/fireorm_sample/blob/master/src/app.module.ts#L55

**Comment by Shashi-GS** [comment]
Oh! you are right. I missed to share the information. can you add "authorization", "content-type", "accept"  properties.

![image](https://user-images.githubusercontent.com/17082069/93991357-d3076900-fda9-11ea-8b4f-9e246bbbedf5.png)

enter accesstoken of your account. Sorry! if you want I can remove that.

**Comment by skalashnyk** [comment]
@Shashi-GS, unfortunately, the problem that you bumped into is well-known one.
@wovalle described the nature of it and possible workarounds: https://github.com/wovalle/fireorm/issues/69#issuecomment-509828334

On our project, we use the next workaround:

```
// foo.entity.ts
@Collection()
export class Foo{
  id: string;
}

// foo.repository.ts
@CustomRepository(Foo) 
export class FooRepository extends BaseFirestoreRepository<Foo>{
  static get instance(){
    return getCustomRepository(Foo) as FooRepository;
  }
  async someCustomMethod(){}
}

// foo.service.ts
export class FooService {
  private get fooRepo(){
      return FooRepository.instance;
  }

  async serviceCall(){
    await this.fooRepo.someCustomMethod();
  }
}
```

**Comment by wovalle** [comment]
Hi @Shashi-GS, unfortunately as @skalashnyk said, you're facing a known limitation.

An easy workaround in your case would be to `require` the custom repository in your app entrypoint or in the files that are being used (which I think you have to do it either way because you have to parse the `getRepository` response to your custom one)

**Comment by skalashnyk** [comment]
@wovalle, It seems, we can do this by passing custom repository class as argument to `getCustomRepository(...)` function.
For example, `typeorm` (that you mentioned above) does it this way - https://typeorm.io/#/custom-repository/custom-repository-extends-standard-repository
In this way, our code will know which exactly class to instantiate. 
What do you think?

**Comment by wovalle** [comment]
That's an interesting take @skalashnyk, I'll check it out

**Comment by wovalle** [comment]
So, by requiring to pass the custom repository class you created to `getCustomRepository` you force the user to import the file that contains the repository (hence running the decorators) and the issue of not having the custom repository in is gone.

This is not an actual fix (still we have to require the files) but using a convention to circumvent the limitation which is brilliant IMHO.

Today, fireorm has 3 methods to get the repositories:

`getRepository`: gets custom repository if defined, gets base repository otherwise
`getBaseRepository`: always return the base repository even if it has a custom repository registered 
`getCustomRepository`: return a the custom repository if registered otherwise return undefined

All of the `getRepository methods require you to pass the Entity constructor as the first parameter (and as a workaround today, you can pass a path as the second parameter but I'm pretty sure that won't be needed after #172). Does it makes sense to extend the first parameter to receive `Entity | YourCustomRepository` and the return type would be either an `IRepository<Entity> | YourCustomRepository`?

This would eliminate the need to parse to [cast the repository returned by getRepository to your custom one](https://fireorm.js.org/#/Custom_Repositories?id=casting) since you'll be specifying the type in the get function.

Would this bring problems down the road? 🤔

**Comment by skalashnyk** [comment]
@wovalle, 
the problem is not only in imports.
Casting (that also requires import) does not solve the problem in current implementation, if we are calling `getCustomRepository(...)` somewhere outside the custom repository class was declared.
Please, look at example, provided by @Shashi-GS in his [comment](https://github.com/wovalle/fireorm/issues/69#issuecomment-697122933):
https://github.com/Shashi-GS/fireorm_sample/blob/master/src/modules/auth/auth.service.ts
As you can see, we imported repository class (`UserDAO`) almost at the top of the file (above all other `fireorm`-related code).

**Comment by skalashnyk** [comment]
>the problem is not only in imports.

I meant, that passing custom repository class to `getCustomRepository(...)` method will allow us to call its (custom repo class) constructor [internally].
And, in this constructor invocation, we will get access to its decorator metadata (entity constructor etc.).
